### DISCUSSION TITLE
"Cross-Resource DependsOn"

### DISCUSSION THREAD

**Discussion by unbis** [fluxcd/flux2#730] [Q&A] [answered]
Hello,

Currently I'm exploring the new Flux a bit and wanted to setup `cert-manager` and a `cluster-issuer` in my test-cluster. For this I decided to use the official Helm template (seeing as the static "vanilla" YAML for each release is generated from this by now). The `cluster-issuer` was supposed to be a simple Kustomization. I realised, however, that `dependsOn` only allows me to specify a resource of its own type, i.e. a `HelmRelease` can only depend on a `HelmRelease`. 

Now to my question, is there a way around this or is it just bad practice to mix two "release-types"?

**Comment by kingdonb** [comment, marked as answer]
DependsOn cannot be cross-resource and this will not be possible, a cross-resource dependency is contraindicated and would not fit with the design, a Kustomization can only depend on another Kustomization. However, this does not mean your use case is impossible.

A `HelmRelease` is a custom resource that is first applied by a `Kustomization`. Any kustomization can absolutely perform health assessments on any custom resources that are created by it, as long as these custom resources report their own status as `HelmRelease` does. You can make a Kustomization that does not report as ready until after a HelmRelease is ready (cert-manager) and another Kustomization that depends on it. So a second kustomization creating your `cluster-issuer` will apply only when the cert-manager release installed and became ready. It is not a Cross-Resource DependsOn, but I think it has the properties you're expecting.

**Reply by unbis** [reply]
Thanks a lot for your answer, I'll do as you suggest and use an "intermediate" Kustomization to achieve my goal.

**Reply by kingdonb** [reply]
Thanks, also please note that `dependsOn` **can be cross-namespace**, but as you suggested, it isn't for use across resource-kinds.

**Reply by squaremo** [reply]
> a cross-resource dependency is contraindicated and would not fit with the design, a Kustomization can only depend on another Kustomization

I would like to hear more about why this is considered the case. It seems like a reasonable expectation to me.

**Reply by kingdonb** [reply]
There is some further expansion on this idea that was written around the same time as my earlier comment: https://github.com/fluxcd/flux2/issues/956#issuecomment-782458039

tl;dr: it isn't in the API because it isn't necessary, thanks to the addition of Health Assessment and `sigs.k8s.io/kustomize/kstatus` which gets you all that and more. I'm not certain of why we still need both, to be frank, except for maybe that `dependsOn` is more portable as it can work with types other than `Kustomization`. In the `toolkit.fluxcd.io` KustomizationSpec, [CrossNamespaceResourceDependency](https://pkg.go.dev/github.com/fluxcd/pkg/runtime/dependency#CrossNamespaceDependencyReference) is a type from our own `fluxcd/pkg` that was leveraged to make [dependsOn](https://toolkit.fluxcd.io/components/kustomize/api/#kustomize.toolkit.fluxcd.io/v1beta1.Kustomization) work in the Kustomization CRD spec.

This type only contains a `namespace` and a `name`, there is no `type` or `kind` column. It is only able to be used for referring from an object of some kind, to another object of the same kind.

The same behavior is available in `HelmRelease`, with similar limitations – `HelmRelease` can `dependOn` any other configured `HelmRelease`, but there is no `healthCheck` struct to wait on arbitrary statuses needed, probably because Helm already has an in-built model of its own for assessing readiness of a release and its subcomponents: `--wait`.

https://toolkit.fluxcd.io/components/kustomize/api/#kustomize.toolkit.fluxcd.io/v1beta1.KustomizationSpec

**Reply by squaremo** [reply]
> it isn't in the API because it isn't necessary, thanks to the addition of Health Assessment and `sigs.k8s.io/kustomize/kstatus` which gets you all that and more

I think "it isn't necessary" works as an argument only if the workaround is as good as a design which works in the way expected.

To recap: the problem to solve is that `cluster-issuer`, which is described by YAMLs, depends on `cert-manager`, which comes as a Helm chart. Naturally you would expect that you could create a (flux) Kustomization for `cluster-issuer`, and a HelmRelease for `cert-manager`, and the first could use `dependsOn` to refer to the first. But this doesn't work because you can only `dependsOn` objects of the same type.

The workaround is to create a (flux) Kustomization that itself creates the HelmRelease for `cert-manager`, and includes the HelmRelease in its health check. Now there's a Kustomization that `cluster-issuer` can depend on, let's call it `cert-manager-helmrelease`. Since the `cert-manager-helmrelease` Kustomization won't be marked as ready before the HelmRelease is marked as ready, the `cluster-issuer` Kustomization won't be applied until (transitively) the `HelmRelease` is ready.

This goes from the situation not supported by the design:

    cluster-issuer (K) --depends--> cert-manager (HR)

to this situation, which will work:

    cluster-issuer (K) --depends--> cluster-issuer-helmrelease (K) --syncs--> cert-manager (HR)

This creates complexity for the user, for no good reason (that I've seen explained). It is hard to grok flux syncing as it is, and this introduces an indirection which isn't part of the user's mental model of their configuration.

If you want a HelmRelease to depend on a Kustomization, you have to do this:

```
helmrelease (HR) --depends--> kustomization-helmrelease (HR) --refers--> kustomization-chart --includes--> kustomization (K)
```

or this:

```
helmrelease-kustomization (K) --syncs--> helmrelease (HR)
                              `--depends--> kustomization (K)
```

I do not think that is as good as a design which works in the expected way.

Are there reasons _not_ to change the design? I can think of one: Each controller would have to know about the other types, to be able to wait for them, and this couples the different bits of API as well as making it difficult to introduce new API objects.

I'm not sure this is true, because I think kstatus can wait for any object given Type- and ObjectMeta. In that case, why not make it possible to depend on any object that kstatus knows about? That addresses the objection above, as well as opening up other possibilities. (Though is it _too_ much blue sky? Possibly.)

**Reply by stefanprodan** [reply]
Having dependencies between Kustomizations and HRs will bring even more pain, not only to users but to us as well. I really don’t want to end up in the same situation as flux1/helm-op cyclical pkg deps. As for users, a HR doesn’t just end up on cluster, it comes from a Kustomization and it’s so easy to get into deadlock situations on each side.

**Reply by squaremo** [reply]
> a HR doesn’t just end up on cluster, it comes from a Kustomization

Ultimately, _everything_ comes from a kustomization (or HelmRelease, in principle). And you can create cycles with just kustomizations as easily as with kustomizations and anything else.

**Reply by squaremo** [reply]
> I really don’t want to end up in the same situation as flux1/helm-op cyclical pkg deps

I don't think this is an inevitable or even difficult to avoid outcome. Firstly, because the APIs are their own modules, so you can depend on them from the controller without them depending on one another; and secondly, because the references are ultimately just strings.

**Reply by birkmose** [reply]
> "intermediate" Kustomization to achieve my goal.

Wonder if anyone would care to share an example of how this is done?

**Reply by squaremo** [reply]
> > "intermediate" Kustomization to achieve my goal.
> 
> Wonder if anyone would care to share an example of how this is done?

I explain the workaround in my comment https://github.com/fluxcd/flux2/discussions/730#discussioncomment-473538.

**Reply by artem-nefedov** [reply]
Here are some very common examples where that would be useful:

- Application from Helm chart using a SOPS-encrypted Secret (which must be deployed from Kustomization, because you don't want to store unencrypted secrets in any object other than Secret).
- Helm chart deploying to a namespace created by Kustomization.

**Reply by SimonStiil** [reply]
This would be awesome if it was explained as part of the documentation of "depends on" for Kustomization.
I was looking up the depends on for both Kustomization and HelmRelease as I have a flow that:
1. Install CRD + secrets through Kustomization 
2. Install Helm chart for an opperator
3. Install Custom Resources through Kustomization 

A flow I'm guessing is quite normal. 
But reading this it makes sense to just have 3 Kustomization that depends on and not think of depends on on the Helm side.
Which is logical when you think of how the HelmRelease are read. But when designing it it was not as logical.

**Reply by webermaster** [reply]
I am trying to implement a version of this dependency right now.  I'm struggling, an example repo would be helpful.

**Reply by kingdonb** [reply]
The repo https://github.com/fluxcd/flux2-kustomize-helm-example is implemented with several examples, see the `production` cluster which has `apps` and `infrastructure` with several inter-dependencies, to put CRDs in order before controllers before applications that depend on the CRD service. Hope this helps!
